ksql-common/src/main/java/io/confluent/ksql/function/UdfIndex.java

```diff
@@ -0,0 +1,322 @@
+    LOG.debug("Current UdfIndex:\n{}", describe());
+
+    final String sqlParamTypes = paramTypes.stream()
+        .map(schema -> schema == null
+            ? null
+            : SchemaUtil.getSchemaTypeAsSqlType(schema.type()))
+        .collect(Collectors.joining(", ", "[", "]"));
+
+    return new KsqlException("Function '" + udfName
+        + "' does not accept parameters of types:" + sqlParamTypes);
+  }
+
+  public Collection<KsqlFunction> values() {
+    return allFunctions.values();
+  }
+
+  private String describe() {
+    final StringBuilder sb = new StringBuilder();
+    sb.append("-ROOT\n");
+    root.describe(sb, 1);
+    return sb.toString();
+  }
+
+  private static final class Node {
+
+    private static final Comparator<KsqlFunction> COMPARE_FUNCTIONS =
+        Comparator.nullsFirst(
+            Comparator
+                .<KsqlFunction, Integer>comparing(fun -> fun.isVarArgs() ? 0 : 1)
+                .thenComparing(fun -> fun.getArguments().size())
+        );
+
+    private final Map<Parameter, Node> children;
+    private KsqlFunction value;
+    private int order = 0;
+
+    private Node() {
+      this.children = new HashMap<>();
+      this.value = null;
+    }
+
+    private void update(final KsqlFunction function, final int order) {
+      if (COMPARE_FUNCTIONS.compare(function, value) > 0) {
+        value = function;
+        this.order = order;
+      }
+    }
+
+    private void describe(final StringBuilder builder, final int indent) {
+      for (final Entry<Parameter, Node> child : children.entrySet()) {
+        if (child.getValue() != this) {
+          builder.append(StringUtils.repeat(' ', indent * 2))
+              .append('-')
+              .append(child.getKey())
+              .append(": ")
+              .append(child.getValue())
+              .append('\n');
+          child.getValue().describe(builder, indent + 1);
+        }
+      }
+    }
+
+    @Override
+    public String toString() {
+      return value != null ? value.getFunctionName() : "EMPTY";
+    }
+
+    public static int compare(final Node a, final Node b) {
+      final int compare = COMPARE_FUNCTIONS.compare(a.value, b.value);
+      return compare == 0 ? -(a.order - b.order) : compare;
+    }
+  }
+
+  /**
+   * A class that represents a parameter, with a schema and whether
+   * or not it is part of a variable argument declaration.
+   */
+  static final class Parameter {
+
+    private static final Map<Type, BiPredicate<Schema, Schema>> CUSTOM_SCHEMA_EQ =
+        ImmutableMap.<Type, BiPredicate<Schema, Schema>>builder()
+            .put(Type.MAP, (a, b) ->
+                Objects.equals(a.keySchema(), b.keySchema())
+                    && Objects.equals(a.valueSchema(), b.valueSchema()))
+            .put(Type.ARRAY, (a, b) -> Objects.equals(a.valueSchema(), b.valueSchema()))
+            .put(Type.STRUCT, (a, b) ->
+                a.fields().isEmpty()
```

Comment:
talked with Hojjat offline on this one - I will revisit this check when we add Struct support and make it more robust. For now it is for backwards compatibility.